keps/sig-node/0034-20190129-pid-limiting.md

```diff
@@ -0,0 +1,186 @@
+            * [Pod to Pod Isolation](#pod-to-pod-isolation)
+            * [Node to Pod Isolation](#node-to-pod-isolation)
+            * [Cgroup Enforcement](#cgroup-enforcement)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+         * [Pod to Pod pid isolation](#pod-to-pod-pid-isolation)
+         * [Node to Pod pid isolation](#node-to-pod-pid-isolation)
+      * [Implementation History](#implementation-history)
+         * [Version 1.10](#version-110)
+         * [Version 1.14](#version-114)
+
+Created by [gh-md-toc](https://github.com/ekalinin/github-markdown-toc)
+
+## Summary
+
+A proposal to enable isolation of pid resources.  It proposes a mechanism to
+enable pod-to-pod PID isolation as well as node-to-pod PID isolation.
+
+## Motivation
+
+Pids are a fundamental resource on Linux hosts.  It is trivial to hit the task
+limit without hitting any other resource limits and cause instability to a host
+machine.
+
+Administrators require mechanisms to ensure that user pods cannot induce pid
+exhaustion that prevents host daemons (runtime, kubelet, etc) from running.  In
+addition, it is important to ensure that pids are limited among pods in order to
+ensure they have limited impact to other workloads on the node.
+
+### Goals
+
+This proposal aims to the following:
+- enable administrator control to provide pod-to-pod pid isolation
+- enable administrator control to provide node-to-pod pid isolation
+
+### Non-Goals
+
+This proposal defers the following:
+- ability for a user to request additional number of pid resources per pod
+
+It is anticipated we will support that via a policy knob that could be
+restricted and/or defaulted via PodSecurityPolicy or LimitRange.  We anticipate
+tracking this work under a separate feature gate `GranularPidLimitsPerPod`.  Any
+defaulting applied to pods today would only be used if the pod had no local pod
+pid limiting policy in future dates.
+
+## Proposal
+
+### User Stories [optional]
+
+1. Administrator can default the number of pids per pod to provide pod-to-pod
+   isolation.
+1. Administrator can reserve a number of allocatable pids to user pods via node
+   allocatable.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Pod to Pod Isolation
+
+To enable pid isolation among pods, the `SupportPodPidsLimit` feature gate is
+defined.
+
+If enabled, the kubelet argument for `pod-max-pids` will write out the
+configured pid limit to the pod level cgroup to the value specified on Linux
+hosts.  If -1, the kubelet will default to the node allocatable pid capacity.
+
+#### Node to Pod Isolation
+
+To enable pid isolation from node to pods, the `SupportNodePidsLimit` feature
+gate is proposed.  If enabled, pid reservations may be supported at the node
+allocatable and eviction manager subsystem configurations.
+
+Node allocatable is a well-established feature concept in the kubelet that
```

Comment:
Link to https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/ for node allocatable.